src/egraph.rs

```diff
@@ -202,6 +216,51 @@ impl<L: Language> EGraph<L> {
         pruned
     }
 
+    pub fn fold_constants(&mut self) -> usize {
+        let mut folded = 0;
+        let mut class_new_node = HashMap::default();
+        // finds a foldable expression for each class.
+        // if any, evalute it and add to map
+        for (id, class) in &self.classes {
+            // scan the nodes for foldable expr.
+            for node in &class.nodes {
+                match node {
+                    Expr::Operator(op, cids) => {
+                        let mut consts = Vec::new();
+                        let mut all_consts = true;
+                        // check if each child is a constant
+                        for cid in cids {
+                            let c = &self.get_eclass(*cid);
+                            if let Some(const_e) = c.get_constant() {
+                                consts.push(const_e);
+                            } else {
+                                all_consts = false;
+                                break;
+                            }
+                        }
+                        // evaluate if all children are constant
+                        if all_consts {
+                            let ne = Expr::Operator(op.clone(), consts);
+                            let v = L::eval(&RecExpr::from(ne));
+                            let const_e = Expr::Constant(v);
+                            class_new_node.insert(*id, const_e);
+                            break;
```

Comment:
What's the point of breaking here? Based on the invariant that you shouldn't have two constants in the same class?

I don't think we should do anything about that in this function. We can assert it later. Plus breaks are ugly :stuck_out_tongue:

Reply:
Asserting only one constant sounds good.

Reply:
To be clear, should I add the assert somewhere or just do nothing for now?

Reply:
See below, I think the right place to assert is while building the id-to-constant hashmap